import/webview/rawwebview.cpp

```diff
@@ -14,31 +14,89 @@
 #include "webengine.h"
 #include "webenginesettings.h"
 
+#include <qmozviewcreator.h>
+
 #include <QtGui/QGuiApplication>
 #include <QtGui/QScreen>
 #include <QtGui/QStyleHints>
 #include <QtGui/QMouseEvent>
 #include <QtQuick/QQuickWindow>
 #include <private/qquickwindow_p.h>
 
+#include <algorithm>
+
 #define CONTENT_ORIENTATION_CHANGED QLatin1String("embed:contentOrientationChanged")
 
 namespace SailfishOS {
 
 namespace WebView {
 
+class ViewCreator : public QMozViewCreator
+{
+public:
+    ViewCreator();
+    ~ViewCreator();
+
+    quint32 createView(const QString &url, const quint32 &parentId) override;
+
+    static std::shared_ptr<ViewCreator> instance();
+
+    std::vector<RawWebView *> views;
+};
+
+ViewCreator::ViewCreator()
+{
+    SailfishOS::WebEngine::instance()->setViewCreator(this);
+}
+
+ViewCreator::~ViewCreator()
+{
+    SailfishOS::WebEngine::instance()->setViewCreator(nullptr);
+}
+
+quint32 ViewCreator::createView(const QString &url, const quint32 &parentId)
+{
+    for (RawWebView *view : views) {
+        if (view->uniqueID() == parentId) {
+            view->openUrlInNewWindow(url);
+            break;
+        }
+    }
+
+    return 0;
+}
+
+std::shared_ptr<ViewCreator> ViewCreator::instance()
+{
+    static std::weak_ptr<ViewCreator> instance;
+
+    std::shared_ptr<ViewCreator> creator = instance.lock();
+    if (!creator) {
+        creator = std::make_shared<ViewCreator>();
+        instance = creator;
+    }
+
+    return creator;
+}
+
+
 RawWebView::RawWebView(QQuickItem *parent)
     : QuickMozView(parent)
+    , m_viewCreator(ViewCreator::instance())
     , m_vkbMargin(0.0)
     , m_acceptTouchEvents(true)
 {
+    m_viewCreator->views.push_back(this);
+
     addMessageListener(CONTENT_ORIENTATION_CHANGED);
 
     connect(this, &QuickMozView::recvAsyncMessage, this, &RawWebView::onAsyncMessage);
+
 }
 
 RawWebView::~RawWebView()
 {
+    m_viewCreator->views.erase(std::find(m_viewCreator->views.begin(), m_viewCreator->views.end(), this));
```

Comment:
If we imagine that we have multiple webviews at the time, this clears a bit too much.

We can handle that separately when it becomes topical so I'd not worry about that now.

Reply:
I'm not sure why I used stl but that's the equivalent of QVector::remove(this).

Reply:
I like the stl.

Reply:
With STL you can repeat `m_viewCreator->views` 3 times in a single statement, isn't that beautiful :)

Reply:
Any updates?

Reply:
On this snippet? No. The constructor appends a pointer to this view to a global list when it's constructed, this line removes that same pointer by finding it's position in the list and erasing the item at that position.

Reply:
Misread this one previously. Sorry.